pvlib/temperature.py

```diff
@@ -821,3 +825,155 @@ def noct_sam(poa_global, temp_air, wind_speed, noct, module_efficiency,
+    H = scipy.linalg.hankel(np.arange(samples_per_window),
+                            np.arange(samples_per_window - 1,
+                                      len(temp_cell_prefixed) - 1))
+    # each row of `subsets` is the values in one window
+    subsets = temp_cell_prefixed[H].T
+
+    # `subsets` now looks like this (for 5-minute data, so 4 samples/window)
+    # where "1." is a stand-in for the actual temperature values
+    # [[nan, nan, nan, nan],
+    #  [nan, nan, nan,  1.],
+    #  [nan, nan,  1.,  1.],
+    #  [nan,  1.,  1.,  1.],
+    #  [ 1.,  1.,  1.,  1.],
+    #  [ 1.,  1.,  1.,  1.],
+    #  [ 1.,  1.,  1.,  1.],
+    #  ...
+
+    # calculate weights for the values in each window
+    if coefficients is not None:
+        a = coefficients
+    else:
+        # values from [1], Table II
+        a = [0.0046, 0.00046, -0.00023, -1.6e-5]
+
+    wind_speed = wind_speed.values
+    p = a[0] + a[1]*wind_speed + a[2]*unit_mass + a[3]*wind_speed*unit_mass
+    # calculate the time lag for each sample in the window, paying attention
+    # to units (seconds for `timedeltas`, minutes for `sample_interval`)
+    timedeltas = np.arange(samples_per_window, 0, -1) * sample_interval * 60
+    weights = np.exp(-p[:, np.newaxis] * timedeltas)
+
+    # Set weights corresponding to the prefix values to zero; otherwise the
+    # denominator of the weighted average below would be wrong.
+    # Weights corresponding to (non-prefix) NaN values must be zero too
+    # for the same reason.
+
+    # Right now `weights` is something like this
+    # (using 5-minute inputs, so 4 samples per window -> 4 values per row):
+    # [[0.0611, 0.1229, 0.2472, 0.4972],
+    #  [0.0611, 0.1229, 0.2472, 0.4972],
+    #  [0.0611, 0.1229, 0.2472, 0.4972],
+    #  [0.0611, 0.1229, 0.2472, 0.4972],
+    #  [0.0611, 0.1229, 0.2472, 0.4972],
+    #  [0.0611, 0.1229, 0.2472, 0.4972],
+    #  [0.0611, 0.1229, 0.2472, 0.4972],
+    #  ...
+
+    # After the next line, the NaNs in `subsets` will be zeros in `weights`,
+    # like this (with more zeros for any NaNs in the input temperature):
+
+    # [[0.    , 0.    , 0.    , 0.    ],
+    #  [0.    , 0.    , 0.    , 0.4972],
+    #  [0.    , 0.    , 0.2472, 0.4972],
+    #  [0.    , 0.1229, 0.2472, 0.4972],
+    #  [0.0611, 0.1229, 0.2472, 0.4972],
+    #  [0.0611, 0.1229, 0.2472, 0.4972],
+    #  [0.0611, 0.1229, 0.2472, 0.4972],
+    #  ...
+
+    weights[np.isnan(subsets)] = 0
+
+    # change the first row of weights from zero to nan -- this is a
+    # trick to prevent div by zero warning when dividing by summed weights
+    weights[0, :] = np.nan
+
+    # finally, take the weighted average of each window:
+    # use np.nansum for numerator to ignore nans in input temperature, but
+    # np.sum for denominator to propagate nans in input wind speed.
+    numerator = np.nansum(subsets * weights, axis=1)
+    denominator = np.sum(weights, axis=1)
+    smoothed = numerator / denominator
```

Comment:
@kanderso-nrel this line emitted a warning despite the comment on lines 967-969 that suggested otherwise. Is this a version-specific issue? My versions are below...

```
INSTALLED VERSIONS
------------------
commit           : 66e3805b8cabe977f40c05259cc3fcf7ead5687d
python           : 3.9.7.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.18363
machine          : AMD64
processor        : Intel64 Family 6 Model 140 Stepping 1, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United States.1252

pandas           : 1.3.5
numpy            : 1.20.3
```

Reply:
Hmm, the only way I can reproduce this is if there is a stretch of NaN in the input lasting longer than 20 minutes -- is that true for your case?  That comment is referring to a div by zero warning cause by a quirk of this implementation (no weights for the very first value), not a div by zero warning caused by nans in the input.

Reply:
Ah, thanks. Yes, I had 38 minutes of NaN temperatures. I filled only the middle point with a valid number and the warning disappeared.